README.md
# deepin-clone
Disk and partition backup/restore tool


CHANGELOG.md
<a name="1.1.1"></a>
## 1.1.1 (2019-03-27)


#### Bug Fixes

*   can not open dim file by application desktop ([d849cd49](https://github.com/linuxdeepin/deepin-clone/commit/d849cd49cd5d365e41ee08ee3798318f7b9d97c0))

#### Features

*   support other file systems, if the executable file corresponding to partclone exists ([12a29530](https://github.com/linuxdeepin/deepin-clone/commit/12a29530fa293b47e5d58aff95a42b00db13aedc))



<a name="1.1.0"></a>
## 1.1.0 (2019-01-10)


#### Bug Fixes

*   "a image" to "an image" ([b9aae895](https://github.com/linuxdeepin/deepin-clone/commit/b9aae89576319419a73257f3f1bbfb02201519a3))



<a name="0.1.2"></a>
### 0.1.2 (2018-11-23)


#### Bug Fixes

*   not upload ts file ([cd981ea9](https://github.com/linuxdeepin/deepin-clone/commit/cd981ea9b7f15461fdeadd666aa91de82f0f57ea))

#### Features

*   add generic name for desktop file ([90871391](https://github.com/linuxdeepin/deepin-clone/commit/9087139142075573ca6e11d402f5a9a7ed7010cb))



<a name="0.1.1"></a>
## 0.1.1 (2018-10-25)


#### Bug Fixes

*   faild on build ([bff840a0](https://github.com/linuxdeepin/deepin-clone/commit/bff840a05ded014ae5ea9ccc170d458799cd89bd))



<a name="0.1.0"></a>
## 0.1.0 (2018-10-25)


#### Bug Fixes

*   the Helper::processExec is blocked if the child process standard error/output hit the limit ([defcefcd](https://github.com/linuxdeepin/deepin-clone/commit/defcefcdeae4673401edea5efa7a07832201f097))



<a name="0.0.9"></a>
## 0.0.9 (2018-07-31)


#### Features

*   add "magic" for mimetype xml file ([1eae4fe1](https://github.com/linuxdeepin/deepin-clone/commit/1eae4fe1be1ae54e2ba8ffe5e50f6000944de7a8))



<a name="0.0.8.5"></a>
### 0.0.8.5 (2018-06-07)


#### Bug Fixes

*   failed on build if the cxx flags contains "-Werror=format-security" ([b9d0c31e](https://github.com/linuxdeepin/deepin-clone/commit/b9d0c31ee185ceda9c2b55c5b4a09277c235145b))



<a name="0.0.8.4"></a>
## 0.0.8.4 (2018-04-24)


#### Bug Fixes

*   failed on build dfm plugin ([4f217e7f](https://github.com/linuxdeepin/deepin-clone/commit/4f217e7f5d1ae151165707c125cc94ce9e8fdbeb))
*   can not show backup file in File Manager ([d670fa60](https://github.com/linuxdeepin/deepin-clone/commit/d670fa600cbfd699db7bd59f745ae148911d8027))



<a name="0.0.8.3"></a>
##  (2018-03-16)


#### Bug Fixes

*   translation of simplified Chinese ([125bcb02](0.0.8.3/commit/125bcb02dad7387684c2bc2e50a2c383d90e6f56))
*   source file is wrong ([b057409e](0.0.8.3/commit/b057409e10f337c1c1e38b8b1b5dce4a62242785))



<a name=""></a>
##  (2018-03-14)


#### Bug Fixes

*   crash in __d_asprintf__ ([07a883c6](0.0.8.2/commit/07a883c6cbaac547d552250ba905275f71844cfe))



<a name=""></a>
##  0.0.8.1 (2018-03-08)


#### Features

*   support xfs ([7a0a92b5](https://github.com/linuxdeepin/deepin-clone/commit/7a0a92b56b09bf920e002294b8c49d7828ff8a6a))

#### Bug Fixes

*   build failed when enable "DISABLE_GUI" on qmake ([f5f64920](https://github.com/linuxdeepin/deepin-clone/commit/f5f64920a33b0834289c2523fb6c74ee719049ee))



<a name=""></a>
##  0.0.8 (2018-02-09)


#### Features

*   add compile test for dfm_plugin project ([4ffbe2be](https://github.com/linuxdeepin/deepin-clone/commit/4ffbe2be7ed26a5a7328b654c4c6130bf35a5959))
*   get QT_SCALE_FACTOR from pkexec user ([69cf6a44](https://github.com/linuxdeepin/deepin-clone/commit/69cf6a44451bbdd1e131e8e0424591462916ec8f))
*   add DDE File Manager plugin support for dim files ([6b907a4e](https://github.com/linuxdeepin/deepin-clone/commit/6b907a4e62ae7f504f2a0d3bea41943e5d8382f5))

#### Bug Fixes

*   crash at DVirtualImageFileIO ([772cdaa8](https://github.com/linuxdeepin/deepin-clone/commit/772cdaa8f16162cd4ba1303fd20d61363f0a12ba))
*   failed build on aarch64 platform ([0541d197](https://github.com/linuxdeepin/deepin-clone/commit/0541d197e9b99328e9cb9a35c16891ee76597add))
*   Adapt lintian ([08cb3e42](https://github.com/linuxdeepin/deepin-clone/commit/08cb3e42198c0e9f390868f3921c77ab668dafa6))
*   chooser file dialog is open by root user ([2e6883b8](https://github.com/linuxdeepin/deepin-clone/commit/2e6883b8c5b038fdd2e4f40cd3b3d2a063cb90df))
*   can not auto start deepin-clone on restart to live system ([76eb0dd2](https://github.com/linuxdeepin/deepin-clone/commit/76eb0dd29b574da782f6465cf81e83512741ae4e))



<a name=""></a>
##  0.0.7.1 (2017-11-28)


#### Bug Fixes

*   not help document for dman ([15425655](https://github.com/linuxdeepin/deepin-clone/commit/1542565580864d34ca1cd3338c654037033da818))



<a name=""></a>
##  0.0.7 (2017-11-16)


#### Bug Fixes

*   can not open url ([28d68b3f](https://github.com/linuxdeepin/deepin-clone/commit/28d68b3f5817a4c9c49dab47487ad0e53d740f28))
*   set the UtilityListItem layout margins to 0 ([56b8f582](https://github.com/linuxdeepin/deepin-clone/commit/56b8f5823f21ef58fbe9218050bc64d5f379d80d))
*   UI for main window ([70d8587d](https://github.com/linuxdeepin/deepin-clone/commit/70d8587d44aeaa475cfc33a858876ac8267520e4))

#### Features

*   use user dbus and user home path ([b8e70e4d](https://github.com/linuxdeepin/deepin-clone/commit/b8e70e4d31430f533a512f0cce524201e038dd49))



<a name=""></a>
##  0.0.6 (2017-11-08)


#### Features

*   add the help documents ([90424d71](https://github.com/linuxdeepin/deepin-clone/commit/90424d7125cce0b80baa1505f13e5cf9e996932b))
*   backup log file to "/var/log" of source system in live system ([2047540b](https://github.com/linuxdeepin/deepin-clone/commit/2047540b3658d9c9e6dd0ea9b7a88b745236e064))
*   disable the window close button on working state ([9b1d6f4e](https://github.com/linuxdeepin/deepin-clone/commit/9b1d6f4e43103289db1a0da8072a71a64f845cd9))
*   add shortcut for show help window ([a3da0c62](https://github.com/linuxdeepin/deepin-clone/commit/a3da0c628edf6056b3e42ed9e17a31ff2a329247))

#### Bug Fixes

*   not install the help document files ([f40fe597](https://github.com/linuxdeepin/deepin-clone/commit/f40fe5972fe50feb778812003d0f09c2412d95e3))

dpartinfo.h
#ifndef DPARTINFO_H
#define DPARTINFO_H

#include "dpartinfo.h"

#include <QList>
#include <QExplicitlySharedDataPointer>

QT_BEGIN_NAMESPACE
class QJsonObject;
QT_END_NAMESPACE

class DPartInfoPrivate;
class DPartInfo
{
public:
    enum Type {
        Invalid,
        Unknow,
        EXT2,
        EXT3,
        EXT4,
        FAT12,
        FAT16,
        FAT32,
        Btrfs,
        F2FS,
        HFS_Plus,
        Minix,
        Nilfs2,
        NTFS,
        Reiser4,
        VFAT
    };

    enum GUIDType {
        InvalidGUID = 0,
        GUIDTypeBegin = 1,
        // None
        Unused_None = GUIDTypeBegin,
        MBR_PS_None = 2,
        EFI_SP_None = 3,
        BIOS_BP_None = 4,
        iFFS_None = 5,
        Sony_BP_None = 6,
        Lenove_BP_None = 7,
        // Windows
        MSR_Win = 8,
        BasicData_Win = 9,
        LDM_Win = 10,
        LDM_DP_Win = 11,
        WRE_Win = 12,
        IBM_GPFS_Win = 13,
        SSP_Win = 14,
        // HP-UX
        DP_HPUX = 15,
        SP_HPUX = 16,
        // Linux
        LFD_Linux = 17,
        RAID_P_Linux = 18,
        RP_x86_Linux = 19,
        RP_x86_64_Linux = 20,
        RP_32bit_ARM_Linux = 21,
        RP_64bit_ARM_Linux = 22,
        SP_Linux = 23,
        LVM_P_Linux = 24,
        Home_P_Linux = 25,
        Srv_P_Linux = 26,
        Plain_DC_P_Linux = 27,
        LUKS_P_Linux = 28,
        Reserved_Linux = 29,
        // FreeBSD
        BP_FreeBSD = 30,
        DP_FreeBSD = 31,
        SP_FreeBSD = 32,
        UFS_P_FreeBSD = 33,
        VVM_P_FreeBSD = 34,
        ZFS_P_FreeBSD = 35,
        // macOS Darwin
        HFS_PLUS_P_Mac = 36,
        UFS_Mac = 37,
        ZFS_Mac = 38,
        RAID_P_Mac = 39,
        RAID_P_Offline_Mac = 40,
        BP_Mac = 41,
        Label_Mac = 42,
        TV_RP_Mac = 43,
        CS_P_Mac = 44,
        SoftRAID_Status_Mac = 45,
        SoftRAID_Scratch_Mac = 46,
        SoftRAID_Volume_Mac = 47,
        SoftRAID_Cache_Mac = 48,
        // Solaris illumos
        BP_Solaris = 49,
        RP_Solaris = 50,
        SP_Solaris = 51,
        Backup_P_Solaris = 52,
        Usr_P_Solaris = ZFS_Mac,
        Var_P_Solaris = 53,
        Home_P_Solaris = 54,
        AS_Solaris = 55,
        Reserved_Solaris = 56,
        // NetBSD
        SP_NetBSD = 57,
        FFS_P_NetBSD = 58,
        LFS_P_NetBSD = 59,
        RAID_P_NetBSD = 60,
        CP_NetBSD = 61,
        EP_NetBSD = 62,
        // ChromeOS
        Kernel_ChromeOS = 63,
        Rootfs_ChromeOS = 64,
        FU_ChromeOS = 65,
        // Haiku
        BFS_Haiku = 66,
        // MidnightBSD
        BP_MidnightBSD = 67,
        DP_MidnightBSD = 68,
        SP_MidnightBSD = 69,
        UFS_P_MidnightBSD = 70,
        VVM_P_MidnightBSD = 71,
        ZFS_P_MidnightBSD = 72,
        // Ceph
        Journal_Ceph = 73,
        DC_EJ_Ceph = 74,
        OSD_Ceph = 75,
        DC_OSD_Ceph = 76,
        DIC_Ceph = 77,
        DC_DIC_Ceph = 78,
        // OpenBSD
        DP_OpenBSD = 79,
        // QNX
        PAFS_QNX = 80,
        // Plan9
        Partition_Plan9 = 81,
        // VMware ESX
        Vmkcore_VMware = 82,
        VMFS_VMware = 83,
        Reserved_VMware = 84,
        // Android-IA
        Bootloader_Android = 85,
        Bottloader2_Android = 86,
        Boot_Android = 87,
        Recovery_Android = 88,
        Misc_Android = 89,
        Metadata_Android = 90,
        System_Android = 91,
        Cache_Android = 92,
        Data_Android = 93,
        Persistent_Android = 94,
        Factory_Android = 95,
        Fastboot_Android = 96,
        OEM_Android = 97,
        // Open Network Install Environment (ONIE)
        Boot_ONIE = 98,
        Config_ONIE = 99,
        // PowerPC
        Boot_PowerPC = 100,
        // freedesktop.org OSes (Linux, etc.)
        SBLC_OSes = 101,
        // Atari TOS
        BD_P_Atari = 102,
        GUIDTypeEnd
    };

    explicit DPartInfo();
    explicit DPartInfo(const DPartInfo &other);
    ~DPartInfo();

    DPartInfo &operator=(const DPartInfo &other);
#ifdef Q_COMPILER_RVALUE_REFS
    DPartInfo &operator=(DPartInfo &&other) Q_DECL_NOTHROW { swap(other); return *this; }
#endif

    void swap(DPartInfo &other);

    QString filePath() const;
    // device name
    QString name() const;
    // internal kernal device name
    QString kname() const;
    int blockSize() const;
    // total size of partition
    // unit: bytes
    qint64 totalSize() const;
    qint64 sizeStart() const;
    qint64 sizeEnd() const;
    qint64 usedSize() const;
    qint64 freeSize() const;
    // file system type of partition
    QString typeName() const;
    Type type() const;
    // mount path of partition
    QString mountPoint() const;
    bool isMounted() const;
    // file system label of partition
    QString label() const;
    // partition label
    QString partLabel() const;
    GUIDType guidType() const;

    void refresh();

    QByteArray toJson() const;

    static GUIDType guidType(const QByteArray &guid);
    static QString guidTypeDescription(GUIDType type);

protected:
    DPartInfo(DPartInfoPrivate *dd);
    static void fromJson(const QJsonObject &root, DPartInfoPrivate *dd);
    static void fromJson(const QByteArray &json, DPartInfoPrivate *dd);

    QExplicitlySharedDataPointer<DPartInfoPrivate> d;

private:
    friend class DPartInfoPrivate;
    friend class DDiskInfo;
    friend class DFileDiskInfoPrivate;
    friend bool operator==(const DPartInfo &first, const DPartInfo &second);
};

inline bool operator==(const DPartInfo &first, const DPartInfo &second)
{
    if (first.d == second.d)
        return true;
    return first.name() == second.name();
}

QT_BEGIN_NAMESPACE
#ifndef QT_NO_DEBUG_STREAM
QDebug operator<<(QDebug deg, const DPartInfo &info);
#endif
QT_END_NAMESPACE

#endif // DPARTINFO_H
